Frameworks/UIKit/UIImage.mm

```diff
@@ -533,666 +462,575 @@ - (instancetype)initWithContentsOfFile:(NSString*)pathAddr {
 
-/**
- @Status Caveat
- @Notes Ignores the UIImageRenderingMode passed in and will always be treated as the orginal ignoring the template.
-*/
-- (UIImage*)imageWithRenderingMode:(UIImageRenderingMode)renderingMode {
-    return [[[UIImage alloc] _initWithCopyOfImage:self WithRenderingMode:renderingMode] autorelease];
+    if (pathStr) {
+      IwFree(pathStr);
+    }
+  }
+  return self;
 }
 
 /**
  @Status Interoperable
 */
-- (instancetype)initWithData:(NSData*)data {
-    return [self initWithData:data scale:1.0];
+- (instancetype)initWithData:(NSData *)data {
+  return [self initWithData:data scale:1.0];
 }
 
 /**
  @Status Interoperable
 */
-- (instancetype)initWithData:(NSData*)data scale:(float)scale {
-    if (data == nil) {
-        TraceVerbose(TAG, L"UIImage: imageWithData, data=nil!");
-        return nil;
-    }
+- (instancetype)initWithData:(NSData *)data scale:(float)scale {
+  RETURN_NULL_IF(!data);
 
-    unsigned char* in = (unsigned char*)[data bytes];
+  if (self = [self init]) {
+    unsigned char *in = (unsigned char *)[data bytes];
     _scale = scale;
-    _imageStretch.origin.x = 0.0f;
-    _imageStretch.origin.y = 0.0f;
-    _imageStretch.size.width = 1.0f;
-    _imageStretch.size.height = 1.0f;
-
-    if ([data length] == 0)
-        return nil;
-
-    // Early out on some file formats so we don't get silly error messages from trying
-    // everything:
 
-    bool loaded = false;
-
-    if ((in[0] == 0xFF && in[1] == 0xD8) || (in[0] == 0xD8 && in[1] == 0xFF)) {
-        m_pImage = CGJPEGImageCreateFromData(data);
-        if (((CGJPEGImageBacking*)m_pImage->Backing()) && ((CGJPEGImageBacking*)m_pImage->Backing())->_orientation) {
-            [self setOrientation:((CGJPEGImageBacking*)m_pImage->Backing())->_orientation];
-        }
-        loaded = true;
+    if ([data length] == 0) {
+      [self release];
+      return nil;
     }
 
-    if (in[0] == 0x89 && in[1] == 'P' && in[2] == 'N' && in[3] == 'G') {
-        m_pImage = CGPNGImageCreateFromData(data);
-        loaded = true;
-    }
-
-    if (in[0] == 'G' && in[1] == 'I' && in[2] == 'F') {
-        loaded = true;
-    }
-
-    if ((in[0] == 'I' && in[1] == 'I') || (in[0] == 'M' && in[1] == 'M')) {
-        loaded = true;
-    }
-
-    if (in[0] == 'B' && in[1] == 'M') {
-        loaded = true;
-    }
-
-    // Fall back on the less common cases:
-    if (!loaded) {
-        TraceVerbose(TAG, L"Unrecognized image");
-        for (int i = 0; i < 64; i++) {
-            TraceVerbose(TAG, L"%02x ", in[i]);
-            if ((i + 1) % 16 == 0)
-                TraceVerbose(TAG, L"");
-        }
-        return nil;
-    }
+    m_pImage = _CGImageGetImageFromData(in, [data length]);
 
     if (!m_pImage) {
-        _deferredImageData.attach([data copy]);
+      TraceVerbose(TAG, L"Unrecognized image");
+      for (int i = 0; i < 64; i++) {
+        TraceVerbose(TAG, L"%02x ", in[i]);
+        if ((i + 1) % 16 == 0)
+          TraceVerbose(TAG, L"");
+      }
```

Comment:
I dont think any of this should be needed.  This method should just wrap the _CGImageGetImageFromData.